tasks/package.js

```diff
@@ -61,29 +61,101 @@ module.exports = function(grunt) {
-    var srcFiles = ['**', '!**/.gitkeep'].concat((config && config.srcFiles && config.srcFiles.length) ? config.srcFiles : '**');
+    var srcFiles = (config && config.srcFiles && config.srcFiles.length) ? config.srcFiles : [];
     var projFiles = (config && config.projFiles && config.projFiles.length) ? config.projFiles : [];
 
     // Look for a package target spec, build destination path.
     var packageName = grunt.option('name') || config.name || 'package';
-    var destPath = grunt.config.get('config.buildPaths.packages') + '/' + packageName;
+
+    // Determine if we are using rsync and vmPath and tarball for performance.
+    var useRsync = grunt.config('config.packages.rsync') || false;
+    // Determine if we are creating a tarball archive.
+    var archive = grunt.config('config.packages.archive') || useRsync;
+
+    // When using rsync, generate to a path that can be mounted in the VM.
+    var vmPath = grunt.config.get('config.packages.vmData') || 'build/vm_data';
+    var destPath = vmPath + '/' + packageName;
+    var finalPath = grunt.config.get('config.buildPaths.packages') + '/' + packageName;
+    if (!useRsync) {
+      // If not using rsync, then directly generate to final path.
+      destPath = finalPath;
+    }
+
     var tasks = [];
+    var cleanPaths = [destPath];
+    if (destPath !== finalPath) {
+      cleanPaths.push(finalPath);
+    }
     grunt.option('package-dest', destPath);
+    grunt.config.set('clean.packages', cleanPaths);
+    tasks.push('clean:packages');
+
+    var excludePaths = ['bower_components', 'node_modules', '.gitkeep'];
+
+    if (useRsync) {
+      // Pull negate conditions from srcFiles into excludePaths.
+      for (var srcIndex = 0; srcIndex < srcFiles.length; srcIndex++) {
+        var item = srcFiles[srcIndex];
+        if (item.substr(0, 1) === '!') {
+          item = item.slice(1);
+          item = item.replace('**', '*');
+          excludePaths.push(item);
+        }
+      }
+
+      // Setup default rsync command and options.
+      var rsync = 'rsync -ahWL --no-l --stats --chmod=Du+rwx ';
+      for (var pathIndex = 0; pathIndex < excludePaths.length; pathIndex++) {
+        rsync = rsync + "--exclude " + excludePaths[pathIndex] + ' ';
+      }
+
+      // Ensure destination exists and sent the rsync.
+      var srcPath = grunt.config('config.buildPaths.html');
+      var destSrc = path.resolve(destPath, grunt.config.get('config.packages.dest.docroot') || '');
+      var rsyncCommand = rsync + srcPath + '/ ' + destSrc + '/';
+
+      grunt.config.set('mkdir.package', {
+        options: {
+          create: [destSrc]
+        }
+      });
+
+      grunt.config('shell.rsync', {
+        command: rsyncCommand
+      });
+    } else {
+      // Use slower copy if rsync is disabled in options.
+      srcFiles.unshift('**');
+      // Add any additional excludePaths to srcFiles.
+      for (var i = 0; i < excludePaths.length; i++) {
+        srcFiles.push('!**/' + excludePaths[i]);
```

Comment:
Something like:

`var _ = require('lodash');` at the top.

```
srcFiles = _.union(srcFiles, _.map(excludePaths, function (item) {
  return '!**/' + item;
}));
```

Union both merges and unique-ifies.

* https://lodash.com/docs/4.16.4#map
* https://lodash.com/docs/4.16.4#union

Looks like we should check on updating lodash.